agent/lib/kontena/websocket/client.rb

```diff
@@ -0,0 +1,767 @@
+  rescue SocketError => exc
+    raise Kontena::Websocket::ConnectError, exc
+  rescue SystemCallError => exc
+    raise Kontena::Websocket::ConnectError, exc
+  end
+
+  # @raise [ArgumentError] Failed adding cert store file/path: ...
+  # @return [OpenSSL::X509::Store]
+  def ssl_cert_store
+    @ssl_cert_store ||= OpenSSL::X509::Store.new.tap do |ssl_cert_store|
+      ca_file = @ssl_params[:ca_file] || ENV['SSL_CERT_FILE']
+      ca_path = @ssl_params[:ca_path] || ENV['SSL_CERT_PATH']
+
+      if ca_file || ca_path
+        if ca_file
+          debug "add cert store file: #{ca_file}"
+
+          begin
+            ssl_cert_store.add_file ca_file
+          rescue OpenSSL::X509::StoreError
+            raise ArgumentError, "Failed adding cert store file: #{ca_file}"
+          end
+        end
+
+        if ca_path
+          debug "add cert store path: #{ca_path}"
+
+          begin
+            # XXX: does not actually raise
+            ssl_cert_store.add_path ca_path
+          rescue OpenSSL::X509::StoreError
+            raise ArgumentError, "Failed adding cert store path: #{ca_path}"
+          end
+        end
+      else
+        debug "use default cert store paths"
+
+        ssl_cert_store.set_default_paths
+      end
+    end
+  end
+
+  # @param ssl_cert [OpenSSL::X509::Certificate]
+  # @raise [Kontena::Websocket::SSLVerifyError]
+  def ssl_verify_cert!(ssl_cert, ssl_cert_chain)
+    unless ssl_cert
+      raise Kontena::Websocket::SSLVerifyError.new(OpenSSL::X509::V_OK, ssl_cert, ssl_cert_chain), "No certificate"
+    end
+
+    ssl_verify_context = OpenSSL::X509::StoreContext.new(ssl_cert_store, ssl_cert, ssl_cert_chain)
+
+    unless ssl_verify_context.verify
+      raise Kontena::Websocket::SSLVerifyError.new(ssl_verify_context.error, ssl_cert, ssl_cert_chain), ssl_verify_context.error_string
+    end
+
+    unless OpenSSL::SSL.verify_certificate_identity(ssl_cert, self.ssl_hostname)
+      raise Kontena::Websocket::SSLVerifyError.new(OpenSSL::X509::V_OK, ssl_cert, ssl_cert_chain), "Subject does not match hostname #{self.ssl_hostname}: #{ssl_cert.subject}"
+    end
+  end
+
+  # @param verify_result [Integer] OpenSSL::SSL::SSLSocket#verify_result
+  # @return [Kontena::Websocket::SSLVerifyError]
+  def ssl_verify_error(verify_result, ssl_cert = nil, ssl_cert_chain = nil)
+    ssl_verify_context = OpenSSL::X509::StoreContext.new(ssl_cert_store)
+    ssl_verify_context.error = verify_result
+
+    Kontena::Websocket::SSLVerifyError.new(ssl_verify_context.error, ssl_cert, ssl_cert_chain, ssl_verify_context.error_string)
+  end
+
+  # @return [OpenSSL::SSL::SSLContext]
+  def ssl_context
+    @ssl_context ||= OpenSSL::SSL::SSLContext.new().tap do |ssl_context|
+      ssl_context.set_params(**@ssl_params,
+        cert_store: self.ssl_cert_store,
+      )
+    end
+  end
+
+  # TODO: connect_deadline to impose a single deadline on the entire process
+  # XXX: specs
```

Comment:
The main remaining TODO is that `ssl_connect` can take longer than the `connect_timeout`, because the same timeout is used for each read/write step of the SSL handshake... needs to be fixed + spec'd to use a combined deadline.